app/jobs/bulk_api_job.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class BulkApiJob < ApplicationJob
+  class Error < StandardError
+    class JobAlreadyRunning < Error
+      def message
+        "User is already running a bulk API job. Users can only run one bulk job at a time."
+      end
+    end
+
+    class MissingUser < Error
+      def message
+        "First argument in perform method must be the User performing the job."
+      end
+    end
+  end
+
+  queue_as :bulk_api_job
+
+  rescue_from(GitHub::Forbidden, Error::JobAlreadyRunning) do
+    user = arguments.first
+    retries = arguments.last[:retries] if arguments.last.is_a? Hash
+    arguments.last[:retries] -= 1 if retries
+    api_wait_time = user.bulk_api_job
+
+    self.class.set(wait_until: api_wait_time).perform_later(*arguments) if retries && retries.positive?
+  end
+
+  before_perform do |job|
+    user = job.arguments.first
+
+    raise Error::MissingUser unless user.is_a? User
```

Comment:
This assumes that `BulkApiJob` will always be a at a `User` level, so we can't use this for running a `organization` level bulk API job, while this is okay for the first iteration, but can we figure out a way to make it use `Organization` or `User`. Or somehow abstract this out?

Reply:
Like adding a `BulkApiJobSupportable` module to a model, and adding methods like `supports_bulk_api_jobs?` for checking support.

Reply:
Correct me if I'm wrong, but in the end the token calling the API and the rate limit will always belong to a User.

Reply:
Yes, thats true, but if it is a `Organization` level job, then like we do now, we can pick a random admin user and use his/her access_token for the job or we can distribute the API calls using tokens of _all_ the admins. This may help us achieve two things:

1. Not exhaust any one particular User's rate limit
2. Faster execution of the `BulkApiJob` as we're using more than one user's token

TLDR; In the future iterations, we could consider passing an organization, then randomize( or distribute) the token selection in order to not exhaust rate limits.

What do you think?

Reply:
Sounds like a great idea 👍 This is something we should definitely add on a future (if not next) iteration